File: algorithms/data-structures/disjoint-sparse-table/python/disjoint_sparse_table.py
import sys


class DisjointSparseTable:
    """Disjoint Sparse Table for O(1) range sum queries."""

    def __init__(self, arr):
        self.n = len(arr)
        # Round up to next power of 2
        self.sz = 1
        self.levels = 0
        while self.sz < self.n:
            self.sz <<= 1
            self.levels += 1
        if self.levels == 0:
            self.levels = 1

        # Pad array
        self.a = arr[:] + [0] * (self.sz - self.n)
        self.table = [[0] * self.sz for _ in range(self.levels)]
        self._build()

    def _build(self):
        if self.sz == 1:
            self.table[0][0] = self.a[0]
            return
        for level in range(self.levels):
            block = 1 << (level + 1)
            half = block >> 1
            for start in range(0, self.sz, block):
                mid = start + half
                # Right half: prefix sums from mid going right
                self.table[level][mid] = self.a[mid]
                for i in range(mid + 1, min(start + block, self.sz)):
                    self.table[level][i] = self.table[level][i - 1] + self.a[i]
                # Left half: suffix sums from mid-1 going left
                if mid - 1 >= start:
                    self.table[level][mid - 1] = self.a[mid - 1]
                    for i in range(mid - 2, start - 1, -1):
                        self.table[level][i] = self.table[level][i + 1] + self.a[i]

    def query(self, l, r):
        """Return sum of arr[l..r] (0-indexed, inclusive)."""
        if l == r:
            return self.a[l]
        # Find the highest differing bit
        level = (l ^ r).bit_length() - 1
        return self.table[level][l] + self.table[level][r]


def disjoint_sparse_table(n, arr, queries):
    dst = DisjointSparseTable(arr)
    return [dst.query(l, r) for l, r in queries]


if __name__ == "__main__":
    data = sys.stdin.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    arr = [int(data[idx + i]) for i in range(n)]; idx += n
    q = int(data[idx]); idx += 1
    queries = []
    for _ in range(q):
        l = int(data[idx]); idx += 1
        r = int(data[idx]); idx += 1
        queries.append((l, r))
    result = disjoint_sparse_table(n, arr, queries)
    print(' '.join(map(str, result)))

File: algorithms/data-structures/disjoint-sparse-table/cpp/disjoint_sparse_table.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class DisjointSparseTable {
    vector<vector<long long>> table;
    vector<long long> a;
    int sz, levels;
public:
    DisjointSparseTable(const vector<int>& arr) {
        int n = arr.size();
        sz = 1; levels = 0;
        while (sz < n) { sz <<= 1; levels++; }
        if (levels == 0) levels = 1;
        a.assign(sz, 0);
        for (int i = 0; i < n; i++) a[i] = arr[i];
        table.assign(levels, vector<long long>(sz, 0));
        build();
    }

    void build() {
        for (int level = 0; level < levels; level++) {
            int block = 1 << (level + 1);
            int half = block >> 1;
            for (int start = 0; start < sz; start += block) {
                int mid = start + half;
                table[level][mid] = a[mid];
                for (int i = mid + 1; i < min(start + block, sz); i++)
                    table[level][i] = table[level][i - 1] + a[i];
                if (mid - 1 >= start) {
                    table[level][mid - 1] = a[mid - 1];
                    for (int i = mid - 2; i >= start; i--)
                        table[level][i] = table[level][i + 1] + a[i];
                }
            }
        }
    }

    long long query(int l, int r) {
        if (l == r) return a[l];
        int level = 31 - __builtin_clz(l ^ r);
        return table[level][l] + table[level][r];
    }
};

int main() {
    int n;
    cin >> n;
    vector<int> arr(n);
    for (int i = 0; i < n; i++) cin >> arr[i];
    DisjointSparseTable dst(arr);
    int q;
    cin >> q;
    for (int i = 0; i < q; i++) {
        int l, r;
        cin >> l >> r;
        if (i) cout << ' ';
        cout << dst.query(l, r);
    }
    cout << endl;
    return 0;
}
